server/src/main/java/org/apache/cassandra/sidecar/livemigration/LiveMigrationInstanceMetadataUtil.java

```diff
@@ -0,0 +1,193 @@
+        }
+        if (instanceMetadata.cdcDir() != null)
+        {
+            dirIndexMap.put(instanceMetadata.cdcDir(), LIVE_MIGRATION_CDC_RAW_DIR_PATH + "/0");
+        }
+        if (instanceMetadata.localSystemDataFileDir() != null)
+        {
+            dirIndexMap.put(instanceMetadata.localSystemDataFileDir(), LIVE_MIGRATION_LOCAL_SYSTEM_DATA_FILE_DIR_PATH + "/0");
+        }
+        for (int i = 0; i < instanceMetadata.dataDirs().size(); i++)
+        {
+            dirIndexMap.put(instanceMetadata.dataDirs().get(i), LIVE_MIGRATION_DATA_FILE_DIR_PATH + "/" + i);
+        }
+        return dirIndexMap;
+    }
+
+    /**
+     * Returns map of directory and set of placeholders that represent the directory.
+     */
+    public static Map<String, Set<String>> dirPlaceHoldersMap(InstanceMetadata instanceMetadata)
+    {
+        Map<String, Set<String>> placeholderMap = new Hashtable<>();
+
+        placeholderMap.put(instanceMetadata.hintsDir(), Collections.singleton(HINTS_DIR_PLACEHOLDER));
+        placeholderMap.put(instanceMetadata.commitlogDir(), Collections.singleton(COMMITLOG_DIR_PLACEHOLDER));
+        if (instanceMetadata.savedCachesDir() != null)
+        {
+            placeholderMap.put(instanceMetadata.savedCachesDir(), Collections.singleton(SAVED_CACHES_DIR_PLACEHOLDER));
+        }
+
+        if (instanceMetadata.cdcDir() != null)
+        {
+            placeholderMap.put(instanceMetadata.cdcDir(), Collections.singleton(CDC_RAW_DIR_PLACEHOLDER));
+        }
+
+        if (instanceMetadata.localSystemDataFileDir() != null)
+        {
+            placeholderMap.put(instanceMetadata.localSystemDataFileDir(),
+                               Collections.singleton(LOCAL_SYSTEM_DATA_FILE_DIR_PLACEHOLDER));
+        }
+
+        List<String> dataDirs = instanceMetadata.dataDirs();
+        for (int i = 0; i < dataDirs.size(); i++)
+        {
+            String dir = dataDirs.get(i);
+            Set<String> placeholders = Collections.unmodifiableSet(
+            new HashSet<>(Arrays.asList(DATA_FILE_DIR_PLACEHOLDER, DATA_FILE_DIR_PLACEHOLDER + "_" + i)));
+            placeholderMap.put(dir, placeholders);
+        }
+
+        return placeholderMap;
+    }
+
+    /**
+     * Converts given live migration file download URL to local path.
+     *
+     * @param fileUrl  Live migration file download URL
+     * @param metadata Cassandra instance metadata
+     * @return local path for given live migration file download URL
+     */
+    public static String localPath(@NotNull String fileUrl,
+                                   @NotNull InstanceMetadata metadata)
+    {
+
+        Map<String, String> urlToLocalDirMap = migrationUrlLocalDirMap(metadata);
+        for (Map.Entry<String, String> entry : urlToLocalDirMap.entrySet())
+        {
+            if (fileUrl.startsWith(entry.getKey()))
+            {
+                assert entry.getValue() != null : "No local path found for url " + fileUrl;
```

Comment:
this statement will only run when assertions are enabled. Do we instead want to throw a runtime exception?
```suggestion
                Objects.requireNonNull(entry.getValue(), () -> "No local path found for url "+ fileUrl);
```

Reply:
done